azurelinuxagent/ga/collect_logs.py

```diff
@@ -324,30 +331,57 @@ def daemon(self):
         metrics = []
-        for cgroup in self.cgroups:
-            metrics.extend(cgroup.get_tracked_metrics(track_throttled_time=True))
+        for controller in self.controllers:
+            metrics.extend(controller.get_tracked_metrics(track_throttled_time=True))
+
+        for metric in metrics:
+            current_max = self.max_recorded_metrics.get(metric.counter)
+            self.max_recorded_metrics[metric.counter] = metric.value if current_max is None else max(current_max, metric.value)
+
         return metrics
 
     def _send_telemetry(self, metrics):
         for metric in metrics:
             report_metric(metric.category, metric.counter, metric.instance, metric.value, log_event=self.__log_metrics)
 
     def _verify_memory_limit(self, metrics):
-        current_usage = 0
+        current_anon_and_swap_usage = 0
+        current_cache_usage = 0
+        memory_throttled_events = 0
         for metric in metrics:
-            if metric.counter == MetricsCounter.TOTAL_MEM_USAGE:
-                current_usage += metric.value
+            if metric.counter == MetricsCounter.ANON_MEM_USAGE:
+                current_anon_and_swap_usage += metric.value
             elif metric.counter == MetricsCounter.SWAP_MEM_USAGE:
-                current_usage += metric.value
-
-        if current_usage > LOGCOLLECTOR_MEMORY_LIMIT:
-            msg = "Log collector memory limit {0} bytes exceeded. The max reported usage is {1} bytes.".format(LOGCOLLECTOR_MEMORY_LIMIT, current_usage)
+                current_anon_and_swap_usage += metric.value
+            elif metric.counter == MetricsCounter.CACHE_MEM_USAGE:
+                current_cache_usage = metric.value
+            elif metric.counter == MetricsCounter.MEM_THROTTLED:
+                memory_throttled_events = metric.value
+
+        mem_limit_exceeded = False
+        if current_anon_and_swap_usage > LOGCOLLECTOR_ANON_MEMORY_LIMIT:
+            mem_limit_exceeded = True
+            msg = "Log collector anon + swap memory limit {0} bytes exceeded. The reported usage is {1} bytes.".format(LOGCOLLECTOR_ANON_MEMORY_LIMIT, current_anon_and_swap_usage)
             logger.info(msg)
-            add_event(
-                name=AGENT_NAME,
-                version=CURRENT_VERSION,
-                op=WALAEventOperation.LogCollection,
-                message=msg)
+            add_event(name=AGENT_NAME, version=CURRENT_VERSION, op=WALAEventOperation.LogCollection, message=msg)
+        if current_cache_usage > LOGCOLLECTOR_CACHE_MEMORY_LIMIT:
+            mem_limit_exceeded = True
+            msg = "Log collector cache memory limit {0} bytes exceeded. The reported usage is {1} bytes.".format(LOGCOLLECTOR_CACHE_MEMORY_LIMIT, current_cache_usage)
+            logger.info(msg)
+            add_event(name=AGENT_NAME, version=CURRENT_VERSION, op=WALAEventOperation.LogCollection, message=msg)
+        if memory_throttled_events > LOGCOLLECTOR_MAX_THROTTLED_EVENTS:
```

Comment:
is logcollector monitor thread running for both v1 and v2? are we checking all three metrics both cgroups? I think we want to check only throttling events limit for v2 but checking others as well as a byproduct? if so we may be killing log collector due other values? should we need to check type as well?

Reply:
We still want to check anon memory and cache memory on v2 and exit if limit is exceeded:

 	• Self-monitor rss/anon memory with 25MB limit and exit if exceeded (v1 and v2)
	• Self-monitor cache/file memory with 150MB limit and exit if exceeded (v1 and v2)
	• Throttle memory if it hits 170MB using memory.high limit (v2 only)

Reply:
I'm confused. do we do self-monitor for v2 as well? I understand for v1 but I feel we are being too hard on this for v2. Memory.high already protecting us from not crossing the limit and process gets throttled(instead oom kill). Is throttling number not enough a good indicator we have to exit the process.

Reply:
Discussed in Teams. Motivation is that anon should not exceed 25MB but the throttling limit only accounts for total mem usage